script.py

```diff
@@ -0,0 +1,129 @@
+import requests
+import csv
+import json
+import os
+import sys
+from urllib.parse import urlparse
+
+OBJECT_SPECS = dict(
+    checkouts=dict(
+        patchset_files=[dict(url=True)],
+    ),
+    builds=dict(
+        input_files=[dict(url=True)],
+        output_files=[dict(url=True)],
+        config_url=True,
+        log_url=True,
+    ),
+    tests=dict(
+        output_files=[dict(url=True)],
+        log_url=True,
+    ),
+)
+
+def extract_fields(spec, data):
+    """
+    Extract values of fields from data according to a specification.
+
+    Args:
+        spec:   The specification of fields to extract the values from.
+        data:   The data to extract the field values from.
+
+    Returns:
+        An array of tuples, where each tuple contains:
+            * A tuple containing dictionary keys leading to the field
+            * The field value
+    """
+    tuples = []
+
+    if isinstance(spec, dict):
+        for k, v in spec.items():
+            if k in data:
+                for t in extract_fields(v, data[k]):
+                    tuples.append(((k,) + t[0], t[1]))
```

Comment:
I don't get this function, but specially this line.

Reply:
I understand it now, but I still feel like its variables names could be improved

Reply:
Let's go with key and value then?